bin/scrapeReference42021.js

```diff
@@ -1,122 +1,123 @@
-			conditions = `<ul>${conditions}</ul>`;
-			formattedTitle = formattedTitle.replace(
-				/<ul>(.*)<\/ul>/,
-				''
-			);
-			formattedTitle = `${formattedTitle}${conditions}`;
-		} else {
-			formattedTitle = `${formattedTitle}`;
-		}
-
-		return {id, title: formattedTitle};
-	};
-
-	const scrapeCriterion = (i, el) => {
-		const element = $(el);
-		const title = element
-			.find('h4')
-			.removeChildren('button')
-			.resolveLinks(options.source)
-			.html()
-			.trim();
-
-		const idMatches = /^Critère (\d+)\.(\d+)/i.exec(title);
-
-		if (idMatches === null) {
-			return null;
-		}
-
-		const id = `${idMatches[1]}.${idMatches[2]}`;
-
-		const level = element.attr('data-level');
-
-		const formattedTitle = title.replace(
-			/^Critère \d+\.\d+\. /i,
-			''
+/**
+ * @param {object} rgaaJsonCriteria
+ * @returns {Array} Array
+ */
+function buildThemes(rgaaJsonCriteria) {
+	try {
+	let themes = [];
+	rgaaJsonCriteria.topics.map((topic) => {
+		themes = [
+			...themes,
+			{
+				'id': `${topic.number}`,
+				'title': `${topic.number}. ${topic.topic}`,
+				'criteria': parseCriteria(topic.criteria, topic.number)
+			}
+		];
+	});
+	console.log(`✅ References list successfully generated.`);
+	return themes;
+	} catch (error) {
+		console.error(
+			`❌ An error occured while generating criteria list: ${error.message}`
 		);
+	}	
+}
 
-		const testElements = $(`li[id^="test-${idMatches[1]}-${idMatches[2]}"]`);
-		const tests = testElements.map(scrapeTest).get();
-
-		return {id, title: formattedTitle, level, tests};
-	};
-
-	const scrapeTheme = (i, el) => {
-		const element = $(el);
-		const title = element
-			.resolveLinks(options.source)
-			.html()
-			.trim();
+/**
+ * @param {object} criterias
+ * @param {string} topicNumber
+ * @returns  {object} Object
+ */
+function parseCriteria(criterias, topicNumber) {
+	return criterias.map((criteria, key) => {
+		return {
+			'id': `${topicNumber}.${criteria.criterium.number}`,
+			'title': `${withoutParagraph(marked(criteria.criterium.title, { renderer }))}`,
+			'level': `${getWcagLevel(criteria.criterium.references[0]?.wcag[0])}`,
+			'tests': getTests(
+				criteria.criterium.tests,
+				criteria.criterium.number,
+				topicNumber
+			)
+		};
+	});
+}
 
-		const idMatches = /^(\d+)/i.exec(title);
+/**
+ * @param {string} wcag
+ * @returns {string} String
+ */
+function getWcagLevel(wcag) {
+	return wcag
+		.split(' ')
+		.pop()
+		.replace(/[{()}]/g, '');
+}
 
-		if (idMatches === null) {
-			return null;
+/**
+ * @param {object} tests
+ * @param {number} criteriumId
+ * @param {string} topicNumber
+ * @return {Array<object>} Array[object] 
+ */
+function getTests(tests, criteriumId, topicNumber) {
+	let testsElements = [];
+	for (const key in tests) {
+		let testsNumbering = `${topicNumber}.${criteriumId}.${key}`;
+		if (Object.hasOwnProperty.call(tests, key)) {
+			testsElements = [
+				...testsElements,
+				{id: testsNumbering, title: marked(objectToList(tests[key]), { renderer })}
+			];
```

Comment:
Comme plus haut, j'ai l'impression qu'un `map` serait plus adapté ici 
```suggestion
return tests.map((test) => ({
	id: `${topicNumber}.${criteriumId}.${key}`, 
	title: marked(objectToList(tests[key]), { renderer })
}));
```

Reply:
Ah oui pardon c'est indexé par des chaînes ! En plus mon code était faux, on avait plus la variable `key`.
Dans ce cas on peut peut-être utiliser `Object.entries` ?
```js
return Object.entries(tests).map(([key, test]) => ({
	id: `${topicNumber}.${criteriumId}.${key}`, 
	title: marked(objectToList(tests[key]), { renderer })
}));
```